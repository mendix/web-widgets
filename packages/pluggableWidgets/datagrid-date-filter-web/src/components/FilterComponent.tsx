import { FilterSelector } from "@mendix/widget-plugin-filter-selector/FilterSelector";
import { useWatchValues } from "@mendix/widget-plugin-hooks/useWatchValues";
import { createElement, CSSProperties, ReactElement, useCallback, useRef, useState } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";
import classNames from "classnames";
import DatePickerComponent from "react-datepicker";
import { useSetInitialConditionEffect } from "../features/initialize";
import { DatePicker, RangeDateValue } from "./DatePicker";
import { useOnResetValueEvent } from "@mendix/widget-plugin-external-events/hooks";

export interface FilterComponentProps {
    adjustable: boolean;
    calendarStartDay?: number;
    className?: string;
    defaultFilter: DefaultFilterEnum;
    defaultValue?: Date;
    defaultStartDate?: Date;
    defaultEndDate?: Date;
    dateFormat?: string;
    locale?: string;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (value: Date | undefined, rangeValues: RangeDateValue, type: DefaultFilterEnum) => void;
    name: string;
    parentChannelName: string | null;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Date | undefined>(props.defaultValue);
    const [rangeValues, setRangeValues] = useState<RangeDateValue>([props.defaultStartDate, props.defaultEndDate]);
    const pickerRef = useRef<DatePickerComponent | null>(null);
    const reset = useCallback(() => {
        setValue(undefined);
        setRangeValues([undefined, undefined]);
    }, []);

    useOnResetValueEvent({
        widgetName: props.name,
        parentChannelName: props.parentChannelName ?? undefined,
        listener: reset
    });

    useWatchValues(
        (_prev, _next) => {
            props.updateFilters?.(value, rangeValues, type);
        },
        [value, rangeValues, type]
    );

    const focusInput = useCallback(() => {
        if (pickerRef.current) {
            pickerRef.current.setFocus();
        }
    }, [pickerRef.current]);

    useSetInitialConditionEffect({
        initialFilterType: type,
        initialFilterValue: value ?? rangeValues,
        updateFilters: props.updateFilters
    });

    return (
        <div
            className={classNames("filter-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            style={props.styles}
        >
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    defaultFilter={props.defaultFilter}
                    id={props.id}
                    onChange={useCallback(
                        type => {
                            setType(prev => {
                                if (prev === type) {
                                    return prev;
                                }
                                focusInput();
                                return type;
                            });
                        },
                        [focusInput]
                    )}
                    options={
                        [
                            { value: "between", label: "Between" },
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" },
                            { value: "empty", label: "Empty" },
                            { value: "notEmpty", label: "Not empty" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                calendarStartDay={props.calendarStartDay}
                dateFormat={props.dateFormat}
                disabledInput={type === "empty" || type === "notEmpty"}
                enableRange={type === "between"}
                locale={props.locale}
                id={props.id}
                placeholder={props.placeholder}
                rangeValues={rangeValues}
                ref={pickerRef}
                screenReaderCalendarCaption={props.screenReaderCalendarCaption}
                screenReaderInputCaption={props.screenReaderInputCaption}
                setRangeValues={setRangeValues}
                setValue={setValue}
                value={value}
            />
        </div>
    );
}
