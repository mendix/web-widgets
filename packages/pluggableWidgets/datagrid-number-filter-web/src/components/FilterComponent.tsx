import { FilterSelector } from "@mendix/widget-kit-web/ui/FilterSelector";
import { debounce } from "@mendix/widget-kit-web/util";
import { Big } from "big.js";
import classNames from "classnames";
import { createElement, CSSProperties, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridNumberFilterProps";

interface FilterComponentProps {
    adjustable: boolean;
    className?: string;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (value: Big | undefined, type: DefaultFilterEnum) => void;
    value?: Big;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Big | undefined>(undefined);
    const [valueInput, setValueInput] = useState<string | undefined>("");
    const inputRef = useRef<HTMLInputElement | null>(null);
    const [[onChange, abortSetValue]] = useState(() => debounce((value?: Big) => setValue(value), props.delay));

    useEffect(() => {
        setValueInput(props.value?.toString() ?? "");
        setValue(props.value);
    }, [props.value]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    // abortSetValue computed just once on mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(() => abortSetValue, []);

    const onFilterTypeChange = useCallback(
        type => {
            setType(prev => {
                if (prev === type) {
                    return prev;
                }
                focusInput();
                return type;
            });
        },
        [focusInput]
    );

    return (
        <div
            className={classNames("filter-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            style={props.styles}
        >
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    id={props.id}
                    defaultFilter={props.defaultFilter}
                    onChange={onFilterTypeChange}
                    options={
                        [
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" },
                            { value: "empty", label: "Empty" },
                            { value: "notEmpty", label: "Not empty" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                disabled={type === "empty" || type === "notEmpty"}
                onChange={e => {
                    const value = e.target.value;
                    if (value && !isNaN(Number(value))) {
                        setValueInput(value);
                        onChange(new Big(Number(value)));
                    } else {
                        setValueInput(value);
                        onChange(undefined);
                    }
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="number"
                value={valueInput}
            />
        </div>
    );
}
