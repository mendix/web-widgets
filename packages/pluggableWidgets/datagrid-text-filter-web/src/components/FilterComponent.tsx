import {
    BaseProps,
    StringHelper,
    FilterList,
    InputWithFilters,
    useInputProps
} from "@mendix/widget-plugin-filtering/controls";
import { createElement } from "react";
import { DefaultFilterEnum } from "../../typings/DatagridTextFilterProps";

const filterDefs: Record<DefaultFilterEnum, string> = {
    contains: "Contains",
    startsWith: "Starts with",
    endsWith: "Ends with",
    greater: "Greater than",
    greaterEqual: "Greater than or equal",
    equal: "Equal",
    notEqual: "Not equal",
    smaller: "Smaller than",
    smallerEqual: "Smaller than or equal",
    empty: "Empty",
    notEmpty: "Not empty"
};

const filters: FilterList<DefaultFilterEnum> = Object.entries(filterDefs).map(
    ([value, label]: [DefaultFilterEnum, string]) => ({
        value,
        label
    })
);

export interface FilterComponentProps extends BaseProps {
    defaultFilter: DefaultFilterEnum;
    value: string | undefined;
    onChange: (value: string | undefined, type: DefaultFilterEnum) => void;
    parentChannelName: string | null;
    changeDelay?: number;
}

export function FilterComponent(props: FilterComponentProps): React.ReactElement {
    const { defaultFilter, value, onChange, parentChannelName, changeDelay, ...baseProps } = props;
    const inputProps = useInputProps({
        name: baseProps.name,
        parentChannelName,
        inputType: "text",
        inputDisabled: filter => filter === "empty" || filter === "notEmpty",
        changeDelay,
        defaultFilter,
        filters,
        value,
        onChange,
        valueHelper: StringHelper
    });
    return <InputWithFilters {...baseProps} {...inputProps} />;
}
