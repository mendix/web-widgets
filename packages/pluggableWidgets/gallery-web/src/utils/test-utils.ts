import { ListExpressionValue, ListWidgetValue, ListActionValue } from "mendix";
import { listExp, listWidget } from "@mendix/widget-plugin-test-utils";
import { ItemHelper } from "../helpers/ItemHelper";

export class WidgetItemBuilder {
    private class: ListExpressionValue<string> | undefined = undefined;
    private content: ListWidgetValue | undefined = undefined;
    private action: ListActionValue | undefined = undefined;

    withItemClass(itemClass: ListExpressionValue<string>): WidgetItemBuilder {
        this.class = itemClass;
        return this;
    }

    withContent(val: ListWidgetValue): WidgetItemBuilder {
        this.content = val;
        return this;
    }

    withAction(val: ListActionValue): WidgetItemBuilder {
        this.action = val;
        return this;
    }

    build(): ItemHelper {
        return new ItemHelper(this.class, this.content, this.action);
    }

    static sample(patch?: (builder: WidgetItemBuilder) => void): ItemHelper {
        const builder = new WidgetItemBuilder()
            .withItemClass(listExp(() => "item-class"))
            .withContent(listWidget(() => "Item content"));

        if (patch) {
            patch(builder);
        }

        return builder.build();
    }
}
