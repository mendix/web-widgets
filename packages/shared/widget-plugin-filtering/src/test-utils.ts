/* eslint-disable @typescript-eslint/explicit-function-return-type */
global.structuredClone = (val: unknown) => JSON.parse(JSON.stringify(val));

export function withRealDates(condition: any) {
    const clone = structuredClone(condition);

    function setRealDates(cond: any) {
        if (cond.name === "or" || cond.name === "and") {
            cond.args.forEach(setRealDates);
            return;
        }
        if (cond.type === "function") {
            setRealDates(cond.arg1);
            setRealDates(cond.arg2);
            return;
        }
        if (cond.type === "literal" && cond.valueType === "DateTime") {
            cond.value = new Date(cond.value);
        }
    }

    setRealDates(clone);

    return clone;
}
