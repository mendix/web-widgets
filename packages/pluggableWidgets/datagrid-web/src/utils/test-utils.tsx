import { createElement } from "react";
import { GUID, ObjectItem } from "mendix";
import { dynamicValue, listAttr, listExp } from "@mendix/widget-plugin-test-utils";
import { WidgetProps } from "../components/Widget";
import { ColumnsType } from "../../typings/DatagridProps";
import { Cell } from "../components/Cell";
import { ColumnId, GridColumn } from "../typings/GridColumn";
import { SelectActionHelper } from "../helpers/SelectActionHelper";
import { FocusTargetController } from "@mendix/widget-plugin-grid/keyboard-navigation/FocusTargetController";
import { PositionController } from "@mendix/widget-plugin-grid/keyboard-navigation/PositionController";
import { VirtualGridLayout } from "@mendix/widget-plugin-grid/keyboard-navigation/VirtualGridLayout";
import { ColumnStore } from "../helpers/state/column/ColumnStore";
import { IColumnParentStore } from "../helpers/state/ColumnGroupStore";

export const column = (header = "Test", patch?: (col: ColumnsType) => void): ColumnsType => {
    const c: ColumnsType = {
        alignment: "left" as const,
        attribute: listAttr(() => "Attr value"),
        dynamicText: listExp(() => "Dynamic text"),
        draggable: false,
        header: dynamicValue(header),
        hidable: "no" as const,
        resizable: false,
        showContentAs: "attribute",
        size: 1,
        sortable: false,
        width: "autoFill" as const,
        wrapText: false,
        visible: dynamicValue(true),
        minWidth: "auto",
        minWidthLimit: 100,
        allowEventPropagation: true
    };

    if (patch) {
        patch(c);
    }

    return c;
};

export function mockSelectionProps(patch?: (props: SelectActionHelper) => SelectActionHelper): SelectActionHelper {
    const props = new SelectActionHelper("None", undefined, "checkbox", false, 5, "clear");

    if (patch) {
        patch(props);
    }

    return props;
}

export function mockGridColumn(c: ColumnsType, index: number): GridColumn {
    const parentStore: IColumnParentStore = {
        isLastVisible(_column: ColumnStore): boolean {
            return false;
        },
        sorting: {
            getDirection(_columnId: ColumnId): ["asc" | "desc", number] | undefined {
                return undefined;
            },
            toggleSort(_columnId: ColumnId): void {
                return undefined;
            }
        }
    };

    return new ColumnStore(index, c, parentStore);
}

export function mockWidgetProps(): WidgetProps<GridColumn, ObjectItem> {
    const id = "dg1";
    const columnsProp = [column("Test")];
    const columns = columnsProp.map((col, index) => mockGridColumn(col, index));

    return {
        CellComponent: Cell,
        className: "test",
        columnsDraggable: false,
        columnsFilterable: false,
        columnsHidable: false,
        columnsResizable: false,
        columnsSortable: false,
        data: [{ id: "123456" as GUID }],
        exporting: false,
        filterRenderer: () => <input type="text" defaultValue="dummy" />,
        hasMoreItems: false,
        headerWrapperRenderer: (_index, header) => header,
        id,
        onExportCancel: jest.fn(),
        page: 1,
        pageSize: 10,
        paginationType: "buttons",
        paging: false,
        pagingPosition: "bottom",
        showPagingButtons: "auto",
        visibleColumns: columns,
        availableColumns: columns,
        columnsSwap: jest.fn(),
        columnsCreateSizeSnapshot: jest.fn(),
        selectionStatus: "unknown",
        setPage: jest.fn(),
        processedRows: 0,
        rowClickable: false,
        selectActionHelper: mockSelectionProps(),
        cellEventsController: { getProps: () => Object.create({}) },
        checkboxEventsController: { getProps: () => Object.create({}) },
        focusController: new FocusTargetController(
            new PositionController(),
            new VirtualGridLayout(1, columns.length, 10)
        )
    };
}
