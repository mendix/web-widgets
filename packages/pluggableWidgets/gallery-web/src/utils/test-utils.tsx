import { createElement } from "react";
import { ClickActionHelper } from "@mendix/widget-plugin-grid/helpers/ClickActionHelper";
import { FocusTargetController } from "@mendix/widget-plugin-grid/keyboard-navigation/FocusTargetController";
import { SelectActionHandler, getColumnAndRowBasedOnIndex } from "@mendix/widget-plugin-grid/selection";
import { listAction, objectItems } from "@mendix/widget-plugin-test-utils";
import { render, RenderResult } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import { ObjectItem } from "mendix";
import { GalleryProps } from "../components/Gallery";
import { ItemEventsController } from "../features/item-interaction/ItemEventsController";
import { ItemHelper } from "../helpers/ItemHelper";
import { ItemHelperBuilder } from "./builders/ItemHelperBuilder";
import { PositionController } from "@mendix/widget-plugin-grid/keyboard-navigation/PositionController";
import { VirtualGridLayout } from "@mendix/widget-plugin-grid/keyboard-navigation/VirtualGridLayout";

export function setup(jsx: React.ReactElement): { user: UserEvent } & RenderResult {
    return {
        user: userEvent.setup(),
        ...render(jsx)
    };
}

export function mockItemHelperWithAction(execute: () => void): ItemHelper {
    return ItemHelperBuilder.sample(builder =>
        builder.withAction(
            listAction(mockAction => {
                const action = mockAction();
                action.execute = execute;
                return action;
            })
        )
    );
}

type Helpers = {
    selectHelper?: SelectActionHandler;
    actionHelper?: ClickActionHelper;
    focusController?: FocusTargetController;
    itemEventsController?: ItemEventsController;
};

type Mocks = {
    onClick?: ReturnType<typeof listAction>;
};

export function mockProps(params: Helpers & Mocks = {}): GalleryProps<ObjectItem> {
    const {
        onClick = undefined,
        selectHelper = new SelectActionHandler("None", undefined),
        actionHelper = new ClickActionHelper("single", onClick),
        focusController = new FocusTargetController(new PositionController(), new VirtualGridLayout(3, 4, 10)),
        itemEventsController = new ItemEventsController(
            item => ({
                item,
                selectionType: selectHelper.selectionType,
                selectionMode: "clear",
                clickTrigger: "single"
            }),
            selectHelper.onSelect,
            selectHelper.onSelectAll,
            actionHelper.onExecuteAction,
            focusController.dispatch,
            selectHelper.onSelectAdjacent,
            3
        )
    } = params;

    const itemHelper = ItemHelperBuilder.sample(builder => {
        if (onClick) {
            builder.withAction(onClick);
        }
    });

    return {
        hasMoreItems: false,
        page: 0,
        pageSize: 10,
        paging: false,
        phoneItems: 2,
        tabletItems: 3,
        desktopItems: 4,
        className: "my-gallery",
        ariaLabelListBox: "Mock props ListBox aria label",
        headerTitle: "Mock props header aria label",
        items: objectItems(3),
        itemHelper,
        selectHelper,
        showHeader: true,
        header: <input />,
        itemEventsController,
        focusController,
        getPosition: (index: number) => getColumnAndRowBasedOnIndex(3, 3, index)
    };
}
