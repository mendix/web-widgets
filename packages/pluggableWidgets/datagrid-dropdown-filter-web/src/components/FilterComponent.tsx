import { useWatchValues } from "@mendix/widget-plugin-hooks/useWatchValues";
import { createElement, CSSProperties, ReactElement, UIEventHandler, useCallback, useEffect } from "react";
import { useSelectState } from "../features/select";
import { EMPTY_OPTION_VALUE, finalizeOptions, parseInitValues } from "../features/setup";
import { Option } from "../utils/types";
import { SelectComponent } from "./SelectComponent";
import { useOnResetValueEvent } from "@mendix/widget-plugin-external-events/hooks";

export interface FilterComponentProps {
    ariaLabel?: string;
    className?: string;
    initialSelected?: string;
    status?: JSX.Element;
    footer?: JSX.Element;
    emptyOptionCaption?: string;
    multiSelect?: boolean;
    id?: string;
    options: Option[];
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (values: Option[]) => void;
    onTriggerClick?: () => void;
    onContentScroll?: UIEventHandler<HTMLUListElement>;
    name: string;
    parentChannelName: string | null;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const multiSelect = !!props.multiSelect;
    const options = finalizeOptions(props.options, { multiSelect, emptyOptionCaption: props.emptyOptionCaption });
    const [state, { toggle, setSelected, reset }] = useSelectState(
        options,
        parseInitValues(props.initialSelected ?? "")
    );

    const onSelect = useCallback((value: string) => {
        if (multiSelect) {
            toggle(value);
        } else {
            setSelected(value === EMPTY_OPTION_VALUE ? [] : [value]);
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useSetInitialConditionEffect(props.updateFilters, options, state.selected);

    useWatchValues(
        (_, [selected]) => {
            props.updateFilters?.(selected.length > 0 ? options.filter(o => selected.includes(o.value)) : []);
        },
        [state.selected]
    );

    useOnResetValueEvent({
        widgetName: props.name,
        parentChannelName: props.parentChannelName ?? undefined,
        listener: reset
    });

    return (
        <SelectComponent
            ariaLabel={props.ariaLabel}
            className={props.className}
            footer={props.footer}
            id={props.id}
            inputValue={state.inputValue}
            multiSelect={multiSelect}
            onSelect={onSelect}
            options={options}
            selected={state.selected}
            status={props.status}
            styles={props.styles}
            tabIndex={props.tabIndex}
            placeholder={props.emptyOptionCaption}
            onContentScroll={props.onContentScroll}
            onTriggerClick={props.onTriggerClick}
        />
    );
}

function useSetInitialConditionEffect(
    updateFilters: ((values: Option[]) => void) | undefined,
    options: Option[],
    selected: string[]
): void {
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(() => updateFilters?.(options.filter(o => selected.includes(o.value))), []);
}
