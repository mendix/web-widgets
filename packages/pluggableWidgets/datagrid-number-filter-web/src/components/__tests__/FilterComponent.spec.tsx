import "@testing-library/jest-dom";
import { createElement } from "react";
import { FilterComponent } from "../FilterComponent";
import { render, fireEvent, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";

jest.useFakeTimers();

describe("Filter component", () => {
    it("renders correctly", () => {
        const dom = render(<FilterComponent adjustable defaultFilter="equal" delay={500} />);

        expect(dom.baseElement).toMatchSnapshot();
    });

    it("renders correctly when not adjustable by user", () => {
        const dom = render(<FilterComponent adjustable={false} defaultFilter="equal" delay={500} />);

        expect(dom.baseElement).toMatchSnapshot();
    });

    it("renders correctly with aria labels", () => {
        const dom = render(
            <FilterComponent
                adjustable
                defaultFilter="equal"
                delay={500}
                screenReaderButtonCaption="my label"
                screenReaderInputCaption="my label"
            />
        );

        expect(dom.baseElement).toMatchSnapshot();
    });

    it("calls updateFilters when value changes", () => {
        const updateFiltersHandler = jest.fn();
        const dom = render(
            <FilterComponent defaultFilter="equal" adjustable delay={500} updateFilters={updateFiltersHandler} />
        );

        act(() => {
            const input = dom.getByRole("spinbutton");
            fireEvent.change(input, { target: { value: "23" } });
        });

        expect(updateFiltersHandler).toBeCalled();
    });

    it("debounces calls for updateFilters when value changes with numbers", () => {
        const updateFiltersHandler = jest.fn();
        render(<FilterComponent defaultFilter="equal" adjustable delay={500} updateFilters={updateFiltersHandler} />);
        const input = screen.getByRole("spinbutton");

        // Initial call with default filter
        expect(updateFiltersHandler).toBeCalledTimes(1);

        act(() => {
            fireEvent.change(input, { target: { value: "0" } });
            jest.advanceTimersByTime(499);
            fireEvent.change(input, { target: { value: "1" } });
            fireEvent.change(input, { target: { value: "2" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(2);

        act(() => {
            fireEvent.change(input, { target: { value: "" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(3);
    });

    it("abort debounced calls when unmounted", () => {
        const updateFiltersHandler = jest.fn();
        const { unmount } = render(
            <FilterComponent defaultFilter="equal" adjustable delay={500} updateFilters={updateFiltersHandler} />
        );
        const input = screen.getByRole("spinbutton");

        expect(updateFiltersHandler).toBeCalledTimes(1);

        act(() => {
            fireEvent.change(input, { target: { value: "0" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(2);

        act(() => {
            fireEvent.change(input, { target: { value: "100" } });
            jest.advanceTimersByTime(250);
            unmount();
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(2);
    });

    it("debounces calls for updateFilters when value changes with decimals", () => {
        const updateFiltersHandler = jest.fn();
        render(<FilterComponent defaultFilter="equal" adjustable delay={500} updateFilters={updateFiltersHandler} />);
        const input = screen.getByRole("spinbutton");

        expect(updateFiltersHandler).toBeCalledTimes(1);

        act(() => {
            fireEvent.change(input, { target: { value: "0.0" } });
            jest.advanceTimersByTime(499);
            fireEvent.change(input, { target: { value: "1.7" } });
            fireEvent.change(input, { target: { value: "4" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(2);

        act(() => {
            fireEvent.change(input, { target: { value: "6.8" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(3);
        expect(input).toHaveValue(6.8);
    });

    it("debounces calls for updateFilters when value changes with invalid input", () => {
        const updateFiltersHandler = jest.fn();
        render(<FilterComponent defaultFilter="equal" adjustable delay={500} updateFilters={updateFiltersHandler} />);
        const input = screen.getByRole("spinbutton");

        expect(updateFiltersHandler).toBeCalledTimes(1);

        act(() => {
            fireEvent.change(input, { target: { value: "t1" } });
            jest.advanceTimersByTime(499);
            fireEvent.change(input, { target: { value: "t2" } });
            fireEvent.change(input, { target: { value: "t3" } });
            jest.advanceTimersByTime(500);
        });

        // Consecutive invalid numbers wont call useState with empty value twice
        // this is why we expect func to be called 1 time
        expect(updateFiltersHandler).toBeCalledTimes(1);

        act(() => {
            fireEvent.change(input, { target: { value: "t4" } });
            jest.advanceTimersByTime(500);
        });

        expect(updateFiltersHandler).toBeCalledTimes(1);
    });
});
