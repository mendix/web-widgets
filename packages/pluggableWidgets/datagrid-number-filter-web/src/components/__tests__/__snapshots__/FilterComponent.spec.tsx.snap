// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders correctly 1`] = `
<body>
  <div>
    <div
      class="filter-container"
      data-focusindex="0"
    >
      <div
        class="filter-selector"
      >
        <div
          class="filter-selector-content"
        >
          <button
            aria-controls="undefined-filter-selectors"
            aria-expanded="false"
            aria-haspopup="true"
            class="btn btn-default filter-selector-button button-icon equal"
          >
             
          </button>
        </div>
      </div>
      <input
        class="form-control filter-input"
        type="number"
        value=""
      />
    </div>
  </div>
</body>
`;

exports[`Filter component renders correctly when not adjustable by user 1`] = `
<body>
  <div>
    <div
      class="filter-container"
      data-focusindex="0"
    >
      <input
        class="form-control"
        type="number"
        value=""
      />
    </div>
  </div>
</body>
`;

exports[`Filter component renders correctly with aria labels 1`] = `
<body>
  <div>
    <div
      class="filter-container"
      data-focusindex="0"
    >
      <div
        class="filter-selector"
      >
        <div
          class="filter-selector-content"
        >
          <button
            aria-controls="undefined-filter-selectors"
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="my label"
            class="btn btn-default filter-selector-button button-icon equal"
          >
             
          </button>
        </div>
      </div>
      <input
        aria-label="my label"
        class="form-control filter-input"
        type="number"
        value=""
      />
    </div>
  </div>
</body>
`;
